src/main/scala/julienrf/json/derived/FieldAdapter.scala

```diff
@@ -0,0 +1,40 @@
+package julienrf.json.derived
+
+import shapeless.Witness
+
+class FieldAdapter[K <: Symbol](w: Witness.Aux[K], a: Adapter) {
+  val name = a(w.value.name)
+}
+
+
+object FieldAdapter {
+  implicit def witnessAdapter[K <: Symbol](implicit w: Witness.Aux[K], a: Adapter): FieldAdapter[K] = {
+    new FieldAdapter[K](w, a)
+  }
+}
+
+trait Adapter extends (String => String)
```

Comment:
In general I tend to avoid extending `=>` for types populated by implicit values, because such values can act as implicit conversions.
(this behaviour should not happen here, though, because the argument and return type are the same, though)